### ISSUE TITLE
Establishing control plane cluster using a VIP fails, when VIP is only for client access

### ISSUE THREAD

**Issue by renukamanavalan** [kubernetes/kubeadm#2153]
## Request for help

I am trying to establish a multi-master control plane cluster behind a Load-balancer.

- Use multiple Kubernetes masters behind the Load-balancer as pool-members
- The load balancer provides a VIP and route the connections from external clients to any active pool-member.
- Create a kubernetes control-plane cluster of all masters, so they have their etcd synced at all times. 
- Now external client may reach any master to get the service.

> **Limitation:**
A client can use the VIP to reach a pool-mamber. But pool-members cannot use the VIP to reach another pool-member. Any request to VIP is looped-back within the same pool-mamber.

Hence if I create first master as declared [here](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#stacked-control-plane-and-etcd-nodes), 
 `sudo kubeadm init --control-plane-endpoint "LOAD_BALANCER_DNS:LOAD_BALANCER_PORT" --upload-certs`

The "kubeadm join" suggested in the spew uses, the same LOAD_BALANCER_DNS.
`kubeadm join LOAD_BALANCER_DNS:LOAD_BALANCER_PORT  --token ...`

When I run the above/suggested command in second master, it uses the given LOAD_BALANCER_DNS:LOAD_BALANCER_PORT, which loops back to itself. Hence fails

Can you please suggest, how to overcome ?

The API server in each master must listen to LOAD_BALANCER_DNS:LOAD_BALANCER_PORT for incoming client requests.
But for internal sync they should use the direct IP address of each other.

Can you please suggest how ?

I am using k8s version v1.18.0.

I highly appreciate any help provided.

**Comment by SataQiu** [comment]
What load balancer are you using? 
How about executing `kubeadm join` before adding the second control-plane into the load balancer pool?

**Comment by renukamanavalan** [comment]
We use "SLB", but it does not change destination address when sending traffic to DIP.

Your proposal could be done. But there is a catch

Say upon joining (`kubeadm join ...`), I add the second master into pool. 

> What would happen if the second master reboot or restart kubelet service ?

It would need to connect to the other master using Load-Balancer VIP, which is not possible.

I have a basic question:
Say there are 3 k8s servers behind a LB. Each of the server need to connect to other two, to be able to sync and as well elect a leader. 

> How is this possible w/o the knowledge of DIP of each server?

**Comment by fabriziopandini** [comment]
> pool-members cannot use the VIP to reach another pool-member

AFAIK an HÀ configuration requires a VIP that the workers should use to communicate to the control plane, and a load balancer that handle the routing from this VIP to the underlying API server instances.
Please note that this doesn't mean that the internal VIP must be the same VIP used outside of the clusters (for clients), but an internal VIP is required. This document provides additional info on the topic https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md

**Comment by renukamanavalan** [comment]
Thank you! Let me read the doc.

**Comment by renukamanavalan** [comment]
This doc is an essential one to help establish a control plane cluster. Can you please mention this in the main doc on 

> [Creating Highly available Clusters with Kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/)

**Comment by renukamanavalan** [comment]
The one limitation [here](https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md#kube-vip) is 

> virtual IP need to be in the same IP subnet

Would it be feasible to create a pool of reachable servers, that each can connect to every other, hence creating full mesh, run k8s master in each and create control plane cluster among them ?

**The main requirement for clustering**: 

> All control plane nodes must/could be in full mesh to be able to sync with each other.

If only we can configure each master with a set of IP-addresses of all other masters and ensure the reachability, they could form a cluster.

Is this feasible?

**Comment by neolit123** [comment]
> This doc is an essential one to help establish a control plane cluster. Can you please mention this in the main doc on

cc @mbert this is something we should do.
potentially replace the line "HAProxy can be used as a load balancer." here:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#first-steps-for-both-methods
with something like "See [this guide](https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md) for examples on setting up a load balancer"

**Comment by renukamanavalan** [comment]
Thanks!

Do we have a solution for k8s masters that can connect to each other but not in the same subnet, to form a control plane cluster?

**Comment by renukamanavalan** [comment]
I realize what I need is 

> [Stacked etcd topology](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/#stacked-etcd-topology)

I would like to get some doc on how to set that up.
Let me close this request.

**Comment by neolit123** [comment]
> I would like to get some doc on how to set that up.

stacked/local etcd is the default mode documented here:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/

external etcd requires extra steps:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/setup-ha-etcd-with-kubeadm/

**Comment by mbert** [comment]
> The one limitation [here](https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md#kube-vip) is
> 
> > virtual IP need to be in the same IP subnet
> 
> Say I have a pool of reachable servers (_need not be in same subnet_), that each can connect to every other, hence creating full mesh.
> Can I run k8s master in each and create control plane cluster among them ?

I would not go as far as stating that generally VIPs can be used only within the same subnet. However most technologies used for implementing VIPs have this constraint due to the way they are implemented. The two examples from the document do have this limitation, but I see no reason why e.g. a virtual address based on DNS names should not be implementable for routed networks, and I would not be too surprised if there were already solutions to this available.

**Comment by renukamanavalan** [comment]
Thanks Martin! Here is my issue with following guidelines as [here](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/).

When I run
`sudo kubeadm init --control-plane-endpoint "LOAD_BALANCER_DNS:LOAD_BALANCER_PORT" --upload-certs ...`

It provides me a join command as 
`sudo kubeadm join LOAD_BALANCER_DNS:LOAD_BALANCER_PORT --token ...`

Now the second master is trying to reach the first using this LOAD_BALANCER VIP, which **fails as this VIP only for external clients** (the worker nodes) to access the servers.

What I need to provide, while joining are two IPs, one the LB-VIP that the API server should listen to and two, the direct-IP of the first master. The **second master should use the direct-IP for any inter communication across nodes**.

per stacked-etcd-topology, the LB-VIP is only used by API servers for requests from worker nodes and use direct-IP for etcd clustering.

I believe, during join process, it learns the direct-IPs of other masters and use it for etcd clustering. But **when it initiates the first join using LB-VIP, is where the failure happens**.

**Comment by mbert** [comment]
> Now the second master is trying to reach the first using this LOAD_BALANCER VIP, which **fails as this VIP only for external clients** (the worker nodes) to access the servers.

Not sure whether I understand you correctly here: is that because you want it that way?
Usually there would be no distinction - your `LOAD_BALANCER_PORT` is simply the entry to your load balancer which then will direct your request  to any of the control plane nodes which are up. Hence your VIP/LB-Port can be used by both, control plane and worker nodes.

> What I need to provide, while joining are two IPs, one the LB-VIP that the API server should listen to and two, the direct-IP of the first master. The **second master should use the direct-IP for any inter communication across nodes**.

Maybe I misunderstand you completely, please correct me if I am wrong. Usually the setup is like this:
- your control plane nodes run the API Server bound to an internal port.
- your load balancers are bound to `LOAD_BALANCER_PORT` through which the API Server is accessed and pass traffic through to the control plane nodes using the above internal port.
- your load balancer instance is selected by where the VIP points to.

This is the setup as I described it in the "HA Considerations" document. 

Now are you having trouble getting this to work, or are you trying to build something completely different?

**Comment by renukamanavalan** [comment]
Here is my setup:

1)  I have SLB Load balancer that provides a VIP with the following characteristics:

    1.a) It accepts traffic on VIP:Port from client/worker-node and pass to a k8s-server/pool-member behind

    1.b) It does **not do NATing**, so k8s server receive traffic with destination IP == VIP
              Implying, the API server running in each control-plane-node should listen to the same VIP::port

    1.c) The return traffic from the k8s server goes directly to client/worker-node (DSR) by passing the Load-Balancer

    1.d) The servers/control-plane-nodes/pool-members cannot talk to each other using this VIP.
              Any access to this VIP address, will loop back to the same server.

Hence I can't use this VIP for join from second master.

From Fabrizio, I learned, that this could be solved using two VIPs -- his response is in the comments [above](https://github.com/kubernetes/kubeadm/issues/2153#issuecomment-634474379)

For the control-plane nodes inter communication, I could use [kube_vip](https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md#kube-vip).

So I use this internal VIP as --control-plane-endpoint  and use the LB-VIP as --apiserver-cert-extra-sans.

Does it sound right ?

`sudo kubeadm init --control-plane-endpoint <Internal VIP>:6443 -apiserver-cert-extra-sans <Loadbalancer VIP>:6443 --upload-certs ...`

**Comment by mbert** [comment]
OK, now I understand your setup better. I agree that a second, internal VIP looks like the way to go. I am putting this into cautious wording however because I am not too familiar with the way the configuration is handled internally, because you would then have to configure all the control plane nodes for the internal and the worker nodes for the external VIP. If the connection information for the worker nodes is kept separate from the connection information for the control plane nodes (which I do not know), then this should work.

**Comment by renukamanavalan** [comment]
Thanks Martin! I am not sure too.

Here is another way. Please share your comments.

1) Add only first master to the Load-Balancer as the only pool-member
2) Run kubeadm init <Load balancer IP>:<port> ...
3) Run "kubeadm join <Load balancer IP>:<port>..." from the second master. As the second mster  is not yet behind  Load balancer, it would succeed
4) Now put the second master under the Load balancer
5) Repeat steps 3 & 4 for every additional master

At this point, **I believe, etcd cluster would be formed using direct-IP.** The API servers would be listening at Load-balancer VIP.

**All good -- for now**

**Do you see any issue in future ?**

It appears to survive system reboot of any master.

**Comment by mbert** [comment]
In the end it all boils down to trying things out. 
Is there any particular reason why you don't want to allow workers, but not the control plane nodes to use the VIP/LB? I mean, if you used the same VIP/LB fo both, things would look far simpler.

**Comment by renukamanavalan** [comment]
"Back end server/pool-members can't use VIP for inter-connectivity" is the limitation, I have to play with.

**Comment by mbert** [comment]
> "Back end server/pool-members can't use VIP for inter-connectivity" is the limitation, I have to play with.

But don't fall worker nodes into the same category?

**Comment by renukamanavalan** [comment]
Worker nodes are clients for the Load balancer.  They are not behind the load balancer. So they are good.

**Comment by renukamanavalan** [comment]
If anyone has any tips/points w.r.t the following two ways, please respond. If you have other better way(s), please suggest.

**My setup:** Described in[ comment above](https://github.com/kubernetes/kubeadm/issues/2153#issuecomment-635496555)

**Way 1**: Described in the [same comment ](https://github.com/kubernetes/kubeadm/issues/2153#issuecomment-635496555)after the setup description.

**Way 2**: Described in another [comment above](https://github.com/kubernetes/kubeadm/issues/2153#issuecomment-635526196).

**Comment by fabriziopandini** [comment]
As far as I understand the problem here, this is not a kubeadm problem, but more a request about how to configure AWS networking. I'm not sure this is the right repo for getting this answer, so my suggestions:
- https://github.com/kubernetes-sigs/cluster-api-provider-aws runs a kubeadm based solution on AWS
- other projects deploying on AWS are 
   - https://github.com/kubernetes-sigs/kubespray
   - https://github.com/kubernetes/kops

All the above projects have working network setups on AWS, so might be that moving the issue there is a good idea...

**Comment by renukamanavalan** [comment]
Sounds right!

We all agree, that when it comes to Networking, there are multiple scenarios and multiple solutions. As a user of kubernetes, what I lacked is some insight w.r.t, how the clustering happened. A basic knowledge on "how" would help us understand the key requirements to meet for successful clustering.

There is very likely such a doc exists. But I am unable to find it yet and still looking for it.

Can you please point to the doc explaining the basics of clustering, especially related to Networking requirements ?

For example, I see the brief on stacked-etcd-clustering. This suits my requirement. But what I presumed (**incorrectly**) is that the clustering happens using, the given control-plane-endpoint address, implying this IP should be available *always* to maintain the cluster.  But what I have learned (_**Thanks to Martin & Fabrizio**_) is that the etcd clustering happens internally using Direct-IP.  The initial join does happen through control-plane-endpoint, but it appears to learn the direct-IP addresses of other nodes and the etcd networking uses that. This means, after the initial join, if the control-plane-endpoint IP would not be available, it still does not affect the clustering.

This is a pretty big item to learn. A very useful piece of data.

I am still *not* 100% confident that what I learned is correct. In fact, if this is correct, by theory, I could use direct IP of any existing master in the cluster, as control-plane-endpoint, when joining a new control-plane-node/master. Because once it joins, it can  learn the direct-IPs of all other control plane nodes and join the cluster. If yes, why do I even need an internal VIP ([kube_vip](https://github.com/kubernetes/kubeadm/blob/master/docs/ha-considerations.md#kube-vip)). This makes me doubt my theory.

In short, I still need the doc.

If there is one, please share. You may also add the link to this doc [here.](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/)

Thank you all for all the support. This doc, would likely help resolve.

**Comment by neolit123** [comment]
> We all agree, that when it comes to Networking, there are multiple scenarios and multiple solutions. As a user of kubernetes, what I lacked is some insight w.r.t, how the clustering happened. A basic knowledge on "how" would help us understand the key requirements to meet for successful clustering.

the short answer is: depends on the topology.
you can setup k8s in a variety of different ways.

kubeadm supports two topologies outlined here:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology/
and the diagram arrows make the k8s node / etcd node connectivity requirements quite obvious.

other solutions out there setup the control-plane differently, for example running kube-apiserver alone on a network host.

if you are looking for a generic guide on how k8s node networking works, it's here:
https://kubernetes.io/docs/concepts/architecture/control-plane-node-communication/
if this is somehow not clear, please file a ticket in k/website and tag it with `/sig network`.

for etcd networking i don't think there is a guide other than this:
https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/
 but what you want is that etcd nodes should be able to reach each other and that depending on the topology kube-apiservers might want to reach some of the etcd nodes.

> But what I have learned (Thanks to Martin & Fabrizio) is that the etcd clustering happens internally using Direct-IP

[1] that is right, and a potential small clarification that can be added for:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/ha-topology
is that kubeadm is using "static boostrap" for etcd nodes:
https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/clustering.md#static

> The initial join does happen through control-plane-endpoint, but it appears to learn the direct-IP addresses of other nodes and the etcd networking uses that. This means, after the initial join, if the control-plane-endpoint IP would not be available, it still does not affect the clustering.

it does not affect the etcd clustering, as etcd in kubeadm is using the static bootstrap and it must know all existing etcd member IPs. but if a CP endpoint (e.g. LB) is no longer available then existing worker nodes will not be able to talk to an api-server. again this is quite obvious from the diagrams.

> I am still not 100% confident that what I learned is correct. In fact, if this is correct, by theory, I could use direct IP of any existing master in the cluster, as control-plane-endpoint, when joining a new control-plane-node/master. 

this pins your nodes to an IP of a particular control-plane's api-server. so if a worker node restarts but that kube-api server is no longer there the node will be stuck without connectivity.

> Because once it joins, it can learn the direct-IPs of all other control plane nodes and join the cluster. If yes, why do I even need an internal VIP (kube_vip). This makes me doubt my theory.

the benefits of a load balancer is that it provides a static endpoint IP/DNS-name that your nodes know and that supposedly always gives them a healthy kube-apiserver.

what you are proposing for bootstrap using direct IPs can work, but this can result in a certificate horror story, so we advise against that:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#considerations-about-apiserver-advertise-address-and-controlplaneendpoint

other than [1], i don't see any potential action items for the kubeadm team.

**Comment by renukamanavalan** [comment]
Thanks a lot for the very detailed  explanation and the links. Let me go through them to help me understand better.

**Comment by neolit123** [comment]
i have sent a couple of minor updates for the HA docs here:
https://github.com/kubernetes/website/pull/21295
